### PR TITLE
Separate the serialization and multicast logic in `Signal`.

### PR DESCRIPTION
**Pull request by andersio**
Break part of the Signal.Core synchronisation logic out, so that we may reuse it for:

1. lowering n-ary operators as event transforms e.g. #573; and
2. (hopefully) simplifying the `combineLatest` and `zip` implementations, which currently replicates part of the `Signal.Core` to deal with recursive terminal events.

### CODE REVIEW

**Comment by mdiep** [comment]
There looks to be a real test failure.

**Comment by mdiep** [comment]
It may be missing this from `Signal.send`:

```swift
			if event.isTerminating {
				self.observers = nil
			}
```

File: Sources/Event.swift [hunk 1]
```diff
@@ -832,3 +837,98 @@ extension Signal.Event where Value == Never {
...
+
+				sendLock.lock()
+				action(event)
+				sendLock.unlock()
+
+				tryToCommitTermination()
+			} else if status.tryTransition(from: .alive, to: .terminated) {
+				// Gracefully ignore any terminal event other than the one
+				// transitions the event stream away from the `alive` status.
+				guard !sendLock.try() else {
```
Comment (mdiep) [inline Sources/Event.swift]: I would invert this `guard`. I think it reads weirdly. The common case (unblocked termination) should be what we're guarding.

**Comment by andersio** [comment]
I have made the observer bag cleaning a `Disposable`, since `self.send` in this case would be running inside the send lock.

**Comment by mdiep** [comment]
That was just my guess as to why the test was failing.

File: Sources/Signal.swift [hunk 2]
```diff
@@ -51,96 +55,55 @@ public final class Signal<Value, Error: Swift.Error> {
...
+			}
 
 			// The generator observer retains the `Signal` core.
-			generator(Observer(action: self.send, interruptsOnDeinit: true), Lifetime(disposable))
+			let sink = Event.makeSynchronizing(self.send, disposable: disposable)
+			let inputObserver = Observer(action: sink, interruptsOnDeinit: true)
+			generator(inputObserver, Lifetime(disposable))
 		}
 
 		private func send(_ event: Event) {
```
Comment (mdiep) [inline Sources/Signal.swift]: I think we should remove this function, inlining it into the `init`. It's no longer safe to call this outside of `sink` above, so leaving it here risks future violations of the signal contract.

**Comment by mdiep** [comment]
In #573, I don't see `makeSynchronizing` reused. Is the intent to make this reusable? Where will it be reused?

**Comment by andersio** [comment]
It is here:
https://github.com/ReactiveCocoa/ReactiveSwift/pull/573/files#diff-ceb9f0d4c919bbf25ca4e3c8b7036b8bR254

**Comment by andersio** [comment]
Not sure if this is the most appropriate way to fit it in. If we look ahead to a HKT future, we perhaps should rely on different types of transforms, instead of enum synchronisation strategies.

**Comment by andersio** [comment]
Hmm, something is off.

**Comment by andersio** [comment]
Fixed in https://github.com/ReactiveCocoa/ReactiveSwift/pull/575/commits/6b252d3866e309be9c8fc53fd36d2a4a44d9678a with refinements in the test case.

**Comment by mdiep** [comment]
I'm not convinced that this is a win.

When I look at this PR by itself, it definitely doesn't seem like an improve its own right. If it makes #573 easier, then it could be worth doing, so I think the question is whether we want to take the 2 PRs.

When I look at #573, it looks very complex. I don't like the look of `SynchronizationStrategy`, and I'm not convinced that it's a net benefit to move all of those operators into event transforms. Maybe it'd be better to keep the separate code paths? Event transforms for simple transforms, `Signal` operators for more complicated ones.

I guess it feels like there are 2 separate types of operators here. Treating them in the same, or mostly-same way, sounds appealing, but isn't necessarily very clean. We might be better off keeping two different approaches. The approach for the more complicated operators doesn't need to be what we have today, but I think something that uses distinct APIs might bring more clarity.

**Pull request closed without merging**